src/frontend/js/page-timing.js

```diff
@@ -129,27 +150,90 @@ PageTiming.recordInactiveTime = function() {
+    console.log('PageTiming.reportEndTime() events so far: ', pageTimingEvents + ', adding ' + PageTiming.eventId);
+    if (pageTimingEvents.indexOf(PageTiming.eventId) === -1) {
+        // Record that this event is being processed.
+        pageTimingEvents.push(PageTiming.eventId);
+        console.log('PageTiming.reportEndTime() added new one: ', pageTimingEvents);
+        localStorage.setItem(PageTiming.localStorageKey, JSON.stringify(pageTimingEvents));
+
+        var message = clusive.djangoMessageQueue.wrapMessage(PageTiming.createEndTimeMessage());
+        console.debug('Reporting page data: ', JSON.stringify(message));
+        var postForm = new FormData();
+        postForm.append('csrfmiddlewaretoken', DJANGO_CSRF_TOKEN);
+        postForm.append('timestamp', new Date().toISOString());
+        postForm.append('messages', JSON.stringify([message]));
+        postForm.append('username', DJANGO_USERNAME);
+        var beaconResult = navigator.sendBeacon('/messagequeue/', postForm);
+
+        // `beaconResult` is `true` or `false` depending on whether the browser
+        // successfully queued the request or not.  (It will fail only if the
+        // size of `postForm`, exceeds the browser's limit, which is unlikely
+        // here).  However, `true` does not mean that the request was sent nor
+        // successfully handled by Clusive.  And, there is no way to tell
+        // if Clusive received the request.  PageTiming.flushEndTimeEvents()
+        // below is a way to remove these event IDs from `localStorage`. 
+        // https://www.w3.org/TR/beacon/#return-values
+        console.debug(`Queued ${PageTiming.eventId} via sendBeacon(): ${beaconResult}`);
     } else {
         console.warn('Pagehide event received, but end time was already recorded');
     }
 };
 
+// For clusive-message-queue's logout handler.
+PageTiming.logoutEndTime = function() {
+    'use strict';
+
+    // Stop event handling via sendBeacon()
+    $(window).off('pagehide', PageTiming.windowEventListener);
+    document.removeEventListener('visibilityChange', PageTiming.docEventListener);
+
+    var pageTimingEvents = PageTiming.getEndTimesProcessed();
+    if (pageTimingEvents.indexOf(PageTiming.eventId) === -1) {
+        // Record that this event is being processed.
+        pageTimingEvents.push(PageTiming.eventId);
+        console.log('PageTiming.reportEndTime() added new one: ', pageTimingEvents);
+        localStorage.setItem(PageTiming.localStorageKey, JSON.stringify(pageTimingEvents));
+
+        // Add the PageEnd event to the messagequeue.
+        clusiveEvents.messageQueue.add(PageTiming.createEndTimeMessage());
+        console.debug(`Queued ${PageTiming.eventId} via clusiveEvents.messageQueue.`);
+    } else {
+        console.warn('Pagehide event received, but end time was already recorded');
+    }
+};
+
+// Utility for retrieving the list of PageEnd times from window.localStorage
+PageTiming.getEndTimesProcessed = function () {
```

Comment:
If we're sending each end time to the server immediately at the moment the page view ends, I do not understand why keeping a list in LocalStorage is necessary any more. That was just a way to keep the info around so that it could be sent later.

Reply:
I thought that another reason for using `localStorage` was to avoid a case where `PageTiming.reportEndTime()` was triggered more than once with the same `eventId`.  There was/is a check for that at the top of `PageTiming.reportEndTime()`.  I kept that check, but with changes to its implementation.  Here is how it used to be done:
> `if (!localStorage['pageEndTime.' + PageTiming.eventId]) { ...`

Using `sendBeacon()` does not necessarily mean the message is sent immediately.  The request is put on some internal browser queue, and the browser will send it in a timely manner (according to the documentation).  Still, since there is no way for the `PageTiming` code to tell if it was successfully sent, then maybe it doesn't actually matter if the same `eventId` is sent multiple times.  But also, empirically speaking, I have not seen the message from the `else` clause in the browser console: 
> `console.warn('Pagehide event received, but end time was already recorded');`

That implies that `PageTiming.reportEndTime()` is only ever called once per `eventId`.  I will happily remove that tracking code and its use of `localStorage`.

Reply:
Duplication of end time reporting is also checked for on the server side, so removing this check seems fine.